### PR TITLE
DOC: Add support for interactive examples with jupyterlite-sphinx

### LINKED ISSUE #19729: DOC: Add interactive examples with jupyterlite-sphinx
### Issue with current documentation:

It would be nice to add some interactivity to SciPy's documentation. This could be done in-browser with  [Jupyterlite](https://jupyterlite.readthedocs.io/en/stable/) running a WASM based kernel like [pyodide](https://pyodide.org/en/stable/). Earlier this year, [Scikit-learn added Jupyterlite buttons](https://github.com/scikit-learn/scikit-learn/pull/25887) to its examples galleries. To see it in action, you can check the "launch lite" button at the bottom of the page [here](https://scikit-learn.org/stable/auto_examples/classification/plot_classifier_comparison.html).

### Idea or request for content:

I've created a Sphinx directive `TryExamples` which was accepted into [jupyterlite-sphinx](https://jupyterlite-sphinx.readthedocs.io/en/latest/), a Sphinx extension that makes it easy to embed Jupyterlite notebooks into Sphinx documentation. You may have seen the [Scientific Python Discourse post](https://discuss.scientific-python.org/t/making-docstring-examples-interactive/812) I made earlier this year. I have an [open PR](https://github.com/jupyterlite/jupyterlite-sphinx/pull/116) which should bring the `TryExamples` directive to a production ready state. I think it's time to start talking about actually deploying this in SciPy's docs.

The extension adds a button to each rendered doctest examples section. Clicking the button will swap the rendered examples in-place with an embedded Jupyterlite notebook. You can see it in action for SciPy 1.11.2's documentation at https://steppi.github.io/scipy.

One thing that needs to be considered before this could be shipped is that we need to take care that the SciPy version being documented is that same as that available on the Jupyterlite kernel used. This means we would either need to restrict the use of the directive to the latest release build which is available in Pyodide, or set up CI builds of SciPy for Pyodide in order to use the directive in dev docs. For now, the use of the directive could be guarded behind an environment variable which is only set for the latest release docs. I'd be curious to know what Scikit-learn does; I wasn't able to figure out from a quick look through their docs.

Scikit-learn gives a disclaimer that the interactive documentation is experimental, may have unexpected behavior, and has long load times. I haven't added anything like this yet, but have thought about adding a little question icon next to each button which displays a tooltip with such a warning.

Feel free to check the examples at https://steppi.github.io/scipy and let me know if there's anything that you think could be improved. The button style, position, and text is easily configurable if you think any of this should be changed.

### Additional context (e.g. screenshots, GIFs)

image

after clicking the button...

image

### PR DESCRIPTION
**Pull request by steppi**
#### Reference issue

Closes #19729 

#### What does this implement/fix?

This PR adds interactive examples using the [TryExamplesDirective](https://jupyterlite-sphinx.readthedocs.io/en/latest/directives/try_examples.html) from [jupyterlite-sphinx](https://jupyterlite-sphinx.readthedocs.io/en/latest/). This directive adds a button which can be pressed to swap doctest Examples sections in-place with embedded Jupyterlite notebooks. These notebooks run locally in the user's browser, adding interactivity to the examples.

#### Additional information

Here is a screenshot of the button 

image

and here is a screenshot showing how things look after the button is pressed

image

I have an experimental deployment up at https://steppi.github.io/scipy/, if anyone wants to try it out before the docs finish building, or if the docs don't build for some reason.

### CODE REVIEW

**Comment by steppi** [comment]
This is set up to use the jupyterlite [pyodide kernel](https://github.com/jupyterlite/pyodide-kernel). SciPy 1.12 is not available for it yet, but when it is this could be backported to 1.12. Before merging, I could disable the interactive examples by adding ignore patterns to the file [try_examples.json](https://github.com/scipy/scipy/blob/cb633dd0d3156cd3feea8241a2edfabf632f3186/doc/source/try_examples.json). The interactive examples can be re-enabled at run time by editing the version of this file which is copied to the build directory of the deployed docs.

Also, the pyodide kernel loads very slowly. Executing the first cell of a notebook can take 30 seconds or more, which may be annoying to users. It may be better to switch to the [xeus-kernel](https://github.com/jupyterlite/xeus) which has faster load times. jupyterlite-sphinx uses xeus for its own docs, https://github.com/jupyterlite/jupyterlite-sphinx/pull/98. @martinRenou, do you have an opinion on the choice of kernel, and an idea of how difficult it would be to make an up-to-date SciPy available in xeus for the SciPy docs?

**Comment by steppi** [comment]
Interesting, it's not working in the circle ci deployment here. It works locally though. Marking as draft. I'll try to figure it out.

**Comment by tupui** [comment]
This is great thank you @steppi! I did a quick test on mobile and it also works there 👍

For Kernel choices and other tweaks, I would do the strict minimum to have the smallest maintenance possible. We tend to customize things a bit too much making it tedious to maintain.

For CircleCI, I think it's not an issue at all as we just use the rendering there to check that new docs render properly.

**Comment by steppi** [comment]
Thanks @tupui. 

> For Kernel choices and other tweaks, I would do the strict minimum to have the smallest maintenance possible. We tend to customize things a bit too much making it tedious to maintain.

OK, cool. I'm not sure yet which would have the lower maintenance burden between xeuss and pyodide, but it makes sense to pick the the one with less.

> For CircleCI, I think it's not an issue at all as we just use the rendering there to check that new docs render properly.

Good to know. I thought maybe it was a sign the extension wasn't working as intended, but it seems to be due to limitations in circle ci's rendered docs.

**Comment by martinRenou** [comment]
> do you have an opinion on the choice of kernel, and an idea of how difficult it would be to make an up-to-date SciPy available in xeus for the SciPy docs?

You can already make a jupyterlite-xeus deployment with scipy 1.11 and xeus-python. I can assist you into doing it, it's quite easy actually, you just need to install jupyterlite-xeus and provide an environment.yml file containing your docs dependencies.

I don't know how hard it would be to get the latest scipy on emscripten-forge (the conda channel used for the WASM environment), we should give it a try.

**Comment by steppi** [comment]
Thanks @martinRenou. It seems like jupyterlite-xeus is the way to go. Yeah, let’s look into how to get the latest SciPy into emscripten-forge.

**Comment by steppi** [comment]
I think this is in good shape now and ready to merge if anyone wants to take a look at it. I've configured it to disable all of the interactive examples, this is just putting the framework in place, so it should be safe to do.

Some more info.

Scipy 1.12 hasn't made it's way into pyodide yet, though the PR for adding it, https://github.com/pyodide/pyodide/pull/4499, has been approved. I mentioned the possibility of using the `jupyterlite-xeus` kernel above, but from my experiments with it, it seems not to make an appreciable difference in load times when importing SciPy. It is also further behind in the supported SciPy versions.

There have been some updates in jupyterlite-sphinx since I last posted an example.

1. The first cell now contains a warning text, warning of potentially large downloads, long load times, ...

image

This idea was borrowed from scikit-learn's jupyterlite interactive examples.

2. As discussed here, https://github.com/scipy/scipy/issues/19729, the interactive examples have been disabled on sufficiently small screens as a heuristic to avoid large downloads on limited data mobile plans.


 You can try it out at the latest deployment at the link below

https://steppi.github.io/scipy/reference/

for a direct link to a specific example see

https://steppi.github.io/scipy/reference/generated/scipy.integrate.quad.html#scipy.integrate.quad

File: doc/source/conf.py
Comment (martinRenou) [inline doc/source/conf.py:405]: Maybe adding a note that those data will then be cached by the browser, so the following openings of notebooks will not download those data again.

Comment (steppi) [reply doc/source/conf.py:405]: Huh. I wonder if it's something to do with the tracking settings I have in firefox, but the caching doesn't work for me. Every time I import scipy in a different examples notebook, it downloads scipy, numpy, and openblas again.

File: doc/source/_static/try_examples.css
Comment (tupui) [inline doc/source/_static/try_examples.css:1]: Can you instead move this to scipy.css? I would prefer if we keep a single file for custom css

File: pyproject.toml
Comment (tupui) [inline pyproject.toml:105]: Please also add these in the environment.yml and also in the doc about the toolchain.

File: doc/source/conf.py
Comment (tupui) [reply doc/source/conf.py:405]: Yeah I would keep this as is. These are "details" and until we remove the experimental note we should not give more guarantee. At most i think we could add a link to the underlying tech.

**Review by tupui — COMMENTED** [review]
Thanks @steppi, I left some minor suggestions. I will find some time next week to try. But we should just get this in 😃

File: doc/source/conf.py
Comment (steppi) [reply doc/source/conf.py:405]: I've been playing around with the network tab in the Firefox web console. I've turned off all tracking protection in firefox, and the caching still doesn't work for me. Every time I execute an import from scipy in a different notebook, scipy, numpy, and openblas get downloaded again. Also, total download size from hitting the button to open the notebook and importing SciPy is actually more like 130MB. I was just adding up the size of the numpy, scipy, and openblas wheels that get downloaded before.

File: pyproject.toml
Comment (steppi) [reply pyproject.toml:105]: Thanks. I missed this.

File: doc/source/_static/try_examples.css
Comment (steppi) [reply doc/source/_static/try_examples.css:1]: Yes, that should work.

File: doc/source/conf.py
Comment (steppi) [reply doc/source/conf.py:405]: Something weird was going on with the 130MB, I haven't witnessed that again

**Comment by steppi** [comment]
Thanks @tupui! I've heard branching for 1.13 could be imminent, so agreed, let's just get this in.

**Review by tupui — APPROVED** [review]
All good to me 👍 thanks again Albert 🚀

**Comment by steppi** [comment]
PR to get SciPy 1.12 into pyodide is merged, https://github.com/pyodide/pyodide/pull/4499. When that makes it into a release I'll submit a follow-up to pin the version

**Comment by martinRenou** [comment]
Super nice thanks!

**Pull request merged**
